tests/RedirectMiddlewareTest.php

```diff
@@ -128,6 +129,41 @@ public function testAddsRefererHeader()
         );
     }
 
+    public function testAddsGuzzleRedirectHeader()
+    {
+        $mock = new MockHandler([
+          new Response(302, ['Location' => 'http://example.com']),
+          new Response(302, ['Location' => 'http://example.com/foo']),
+          new Response(302, ['Location' => 'http://example.com/bar']),
+          new Response(200)
+        ]);
+
+        $stack = new HandlerStack($mock);
+        $stack->push(Middleware::redirect());
+        $handler = $stack->resolve();
+        $request = new Request('GET', 'http://example.com?a=b');
+        $promise = $handler($request, [
+            'allow_redirects' => [
+                  'track_redirects' => true,
+            ],
+        ]);
+        /** @var ResponseInterface $response */
+        $response = $promise->wait(TRUE);
+
+        $this->assertEquals(
+          'http://example.com?a=b',
+          $response->getHeaderLine('X-Guzzle-Redirect-0')
```

Comment:
Do we want the headers distinctly named or should they share a name?  Distinctly named is easier if you know what you're looking for.  Shared name means they come back as a collection, which is better if you don't know exactly how many you're working with.

I think I'd marginally prefer a shared header, but defer to @mtdowling